open-grant-proposals/open-proposal-dynamically-backend.md

```diff
@@ -0,0 +1,46 @@
+Name of Project: Dynamically configurable storage backend
+Proposal Category: devtools-libraries
+Proposer: ipfsmainoffical
+Do you agree to open source all work you do on behalf of this RFP and dual-license under MIT and APACHE2 licenses?: Yes
+Project Description
+During the actual use of Filecoin, miners are very unlikely to install all hard drives at once, so they will frequently add new hard drives or replace old hard drives for a variety of reasons. Current Filecoin has a hard time supporting this. 
+We want to develop a disk management backend that supports dynamic configuration to help miners use a variety of hard drives at any time. 
+Our storage backend supports multiple directories and can either dynamically configure new directories or delete old ones. The storage backend uses these directories through user-defined strategies like “less free space first”, “more free space first”, “space reserve”, etc.  
+    1) Maybe user can set a new directory via config or delete the old one.
+    2) Develop a monitoring module for the disk, which has been damaged or removed.
+    3) Develop multiple directory access interface in the rust-fil-sector-builder library, and maintain the correspondence between the sector and the directory.
+    4) Develop different disk usage strategies to let users decide how to use their disks, Is each disk balanced, or one by one. 
+    5) Users can choose to be invisible to these processes, no need to pay attention to storage details.
+
+Value
+Mask the complexity of hard drives operations for ordinary miners. Allow miners to add or remove hard drives according to their needs, instead of having all the hardware ready and professionally operated like raids. 
+Lack of this function will probably prevent miners from interest or use Filecoin network.
+The difficulty should be how to do more in line with the miner’s needs, and the actual test will encounter a variety of problems with the hard drives. 
+Deliverables
```

Comment:
Can you describe these deliverables in greater detail? I'd expect a few more things, including documentation of the project.

Reply:
We will eventually provide:
1) A sectorbuilder library that supports multiple directories.
2) Provide an interface to configure the directory (add, delete) and configure the policy.
3) Provide relevant design documentation, api documentation, and test reports.

Some descriptions are as follows:
1. We believe that storage is a complex issue for the average user.
2. Provide a sectorbuilder that supports multiple directories.
3. It will be implemented in disk_backed_storage.rs.
4. Plan to support multiple strategies for using directories, sequential use, balanced use, and more.
5. raid is complex and uneconomical for the average user.
6. If the user wants to balance the risk of his storage, he can choose to use multiple directories in a balanced manner, so that when one directory is damaged, it does not affect the other.
7. If the user wants to add his storage space one by one, he can choose to add the directory slowly.
8. This approach should be simpler than aggregating file systems and does not introduce a new underlying file system instability and maintenance issues that we believe should be more suitable for mining.